### ISSUE TITLE
Question about torrent_status fields

### ISSUE THREAD

**Issue by elgatito** [arvidn/libtorrent#4667]
Hello,

I am usually setting file priority to `> 0` for only specific file in a multi-file torrents.
Never noticed, if `torrent_status->progress` is properly calculating progress of only selected files/pieces.

Now looking at `torrent_status->progress` calculation then came to `torrent->bytes_done`.
https://github.com/arvidn/libtorrent/blob/RC_1_1/src/torrent.cpp#L4003

Looks like total size is used as a base?
```
st.total_wanted = m_torrent_file->total_size();
```

And later there is a code part, which I don't understand (the logic)

```
                int num_filtered_pieces = m_picker->num_filtered()
			+ m_picker->num_have_filtered();
		int last_piece_index = m_torrent_file->num_pieces() - 1;
		if (m_picker->piece_priority(last_piece_index) == 0)
		{
			st.total_wanted -= m_torrent_file->piece_size(last_piece_index);
			TORRENT_ASSERT(st.total_wanted >= 0);
			--num_filtered_pieces;
		}
		st.total_wanted -= boost::int64_t(num_filtered_pieces) * piece_size;
```

And then later actual calculation of `done` is made:

```
					int deduction = (std::min)(p.length, piece_size - p.start);
					bool done = m_picker->has_piece_passed(j);
					bool wanted = m_picker->piece_priority(j) > 0;
					if (done) st.total_done -= deduction;
					if (wanted) st.total_wanted -= deduction;
					if (wanted && done) st.total_wanted_done -= deduction;
```

Why am I looking at it, if that when I add a torrent file, then select some file for downloading, then prioritize some pieces inside that file. But progress looks like expecting more bytes than that file size.

**Comment by arvidn** [comment]
I think `wanted` and `done` are both quantized to piece sizes. If you select a torrent whose size is not divisible by the piece size, the actual number of bytes you'll download is the file size rounded up to a piece size (somewhat simplified, it depends a bit on if the start of the file is aligned to a piece as well).

Basically, any piece that overlaps the file will be downloaded.

I didn't look closely at the code you reference, but does this make your observations make sense?

**Comment by elgatito** [comment]
@arvidn few users reported that they select one file from multi-file torrent, it gets to 100% downloaded. 
After program is restarted and torrent files are re-added to libtorrent session with resume data, it shows that it is not 100% downloaded. 
That is why I try to understand how progress works, maybe I read wrong fields to calculate the progress.

I will add more debugging information to understand what happens there.

PS: It's RC_1_1 branch.

**Comment by arvidn** [comment]
One thing to look for is if the same number of downloaded pieces reported after restarting with resume data?


I assume no files (or partfiles) were removed when restarting too, right?

**Comment by elgatito** [comment]
@arvidn I am in the progress of adding proper `torrent_status` logging to see what happens.
I have a feeling that I made some errors with setting files priority after program restart.

I thought `total_wanted` and `total_wanted_done` reflects bytes from pieces with `priority > 0`. 
What fields of `torrent_status` can give that information even after totally downloaded torrent is added to the session?

**Comment by elgatito** [comment]
@arvidn one more, is it possible to get information about trackers, their peers count, last update? 
And information about peers for a torrent, speed, etc?

**Comment by arvidn** [comment]
> I thought total_wanted and total_wanted_done reflects bytes from pieces with priority > 0

I think so to. Do you have observations suggesting that's not the case?

> What fields of torrent_status can give that information even after totally downloaded torrent is added to the session?

I don't quite get the condition. You mean if the torrent you added didn't download anything, will it still have `done` bytes? Yes, that's why I called them `done` rather than `downloaded`.

**Comment by elgatito** [comment]
@arvidn I have added proper session logging, and I was right that I've messes with file priorities after program restart, but I have found interesting behavior.

I am adding a torrent with setting all file priorities to 0. 
Setting some file priority to 1. 
It starts to download.

Then I restart the program, re-add the torrent with resume data, also with all files priorities set to `0`.
I see that file priorities, returned by `torrent_status`, are showing `0` everywhere. 
But, libtorrent is still downloading something, while `total_wanted` is also `0`.

**Comment by arvidn** [comment]
> But, libtorrent is still downloading something

Do you mean it's downloading torrent payload? or do you mean there are bytes being sent and received, by, say the DHT or trackers?

**Comment by elgatito** [comment]
@arvidn 
> Do you mean it's downloading torrent payload?

No, regular download, megabytes per second. 
Like priorities are cached somewhere.

**Comment by arvidn** [comment]
what do piece priorities show?
It's fundamentally the piece priorities that count, changing file priorities is short-hand for setting the piece priorities. Perhaps there's an issue where file priorities set via `add_torrent_params` fail to update the piece priorities

**Comment by elgatito** [comment]
@arvidn Here is the output for a torrent, that has multiple files, initially added with 0 priorities, set priority for one file, wait a little and take log.

```
    Files (Priority):
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E01.720p.WEBRip.Profix.Media.mkv (929 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E02.720p.WEBRip.Profix.Media.mkv (871 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E03.720p.WEBRip.Profix.Media.mkv (807 MB): 1 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E04.720p.WEBRip.Profix.Media.mkv (853 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E05.720p.WEBRip.Profix.Media.mkv (849 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E06.720p.WEBRip.Profix.Media.mkv (885 MB): - 

    Pieces:
    ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------++++++++++++++++++1++++11+11111111111111111111111111111111111111111111111111111111111111111111111---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Then I restart the program, add torrent with resume data, and not setting any priorities.

Log after:
```
    Size:
        Total:                  5.2 GB 
        Done:                   482 MB (9.29%) 
        Wanted:                 814 MB 
        Wanted done:            482 MB (59.29%) 

    Files (Priority):
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E01.720p.WEBRip.Profix.Media.mkv (929 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E02.720p.WEBRip.Profix.Media.mkv (871 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E03.720p.WEBRip.Profix.Media.mkv (807 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E04.720p.WEBRip.Profix.Media.mkv (853 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E05.720p.WEBRip.Profix.Media.mkv (849 MB): - 
        Into.The.Night.S01.720p.WEBRip.Profix.Media/Into.The.Night.S01E06.720p.WEBRip.Profix.Media.mkv (885 MB): - 

    Pieces:
    ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------++++++++++++++++++++++++++++++++++++++1++1++1++++++++1+111111111111111111111111111111111111111111---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
    
```

Priorities returned as:
- = no priority
+ = piece finished
number = priority

**Comment by arvidn** [comment]
> Then I restart the program, add torrent with resume data, and not setting any priorities.

You say "with resume data". What does the resume data say about file- and piece priorities?
i.e. are you sure the *saved* resume data from your previous session is correct?
For instance, do you save resume data *after* setting piece priorities to 0?

You say you don't set any priorities. Do you mean you don't modify the `add_torrent_params::file_priorities` nor `add_torrent_params::piece_priorities` *after* you `read_resume_data()`?

**Comment by elgatito** [comment]
@arvidn 

Here is the resume data before restarting.
[db5cecb43153a922228af7876862df31d71c67ea.txt](https://github.com/arvidn/libtorrent/files/4696990/db5cecb43153a922228af7876862df31d71c67ea.txt)

Resume data is synced every 10 seconds.
So it is definitely written after priorities has been set.

> You say you don't set any priorities. Do you mean you don't modify the add_torrent_params::file_priorities nor add_torrent_params::piece_priorities after you read_resume_data()?

`add_torrent_params::file_priorities` torrent is always added with file priorities set to `0` for index `0..500`. Only after torrent is added - I set file priorities.

**Comment by arvidn** [comment]
those fields are populated by `read_resume_data()`. Are you asserting that they are all set to zero once you've loaded the resume data, or are you assuming they are zero because you didn't set them?

**Comment by arvidn** [comment]
Here's what I see in that resume data:

```
'piece_priority':
'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

and

```
'file_priority': [ 0, 0, 0, 1, 0, 0 ]
```

So when you load that resume data back and start the torrent, it makes
sense that you would keep downloading that file, with index 3.

**Comment by elgatito** [comment]
@arvidn 
> those fields are populated by read_resume_data(). Are you asserting that they are all set to zero once you've loaded the resume data, or are you assuming they are zero because you didn't set them?

I don't know really. 
I see that I read file priorities from `torrent_status` and it says no files have priorities.

I assume `add_torrent_params->file_priorities` are overwriting priorities from resume data?
But then again piece priorities are not being reset.

I will probably add a logic to not set file_priorities when I have resume data.

**Comment by arvidn** [comment]
> I see that I read file priorities from torrent_status and it says no files have priorities.

Oh, I see. So, you're seeing file priorities set to 0, but file priorities set to non-zero. I assume by "no priority" you mean priority 0, i.e. do-not-download.

> I assume add_torrent_params->file_priorities are overwriting priorities from resume data?

add_torrent_params::file_priorities *is* the resume data. It's what you get back from `read_resume_data()`

**Comment by arvidn** [comment]
If you can think if things that can be clarified in the documentation about this, I'd appreciate a PR, or comments

**Comment by arvidn** [comment]
what was the issue? no bug in libtorrent?

**Comment by elgatito** [comment]
@arvidn Not a bug in libtorrent, more a clarification/understanding issue. 
Found few bugs in my management of file priorities.
And later needed to figure out how using both `resume_data` and `file_priorities` inside of `add_torrent_params` work together.

The only thing that is not clear, if `file_priorities` are overwriting those that are in `resume_data`, then they probably should update piece priorities as well. Because `resume_data` has both file and piece priorities information. But if I want to explicitly add torrent without starting download - I set file priorities, but piece priorities are still there and download is continued, even when showing file is not needed for download.

**Comment by arvidn** [comment]
can you think of a clarification to the documentation? Or do you think the behavior should change? Say, where either file priorities or piece priorities are saved in the resume file, but never both.

**Comment by elgatito** [comment]
@arvidn I would probably expect another logic:
- `add_torrent_params->resume_data` to set `file_priorities` and `piece_priority`
- `add_torrent_params->file_priorities` to set `file_priorities` and `piece_priority`
- `add_torrent_params->piece_priorities` to set `piece_priority`

In that specific order, so that any explicit priority setting is taking advantage over "under the hood" resume data.

But that is my opinion. 
Still it's not a problem with current implementation, but it needs more information about setting initial priorities with resume data.

**Comment by arvidn** [comment]
`add_torrent_params->resume_data` is deprecated. It only exists for backwards compatibility and will be removed in a future release.

When loading resume data (with the current API) you get an `add_torrent_params` object back, filled in with the resume data.
